sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,571 @@
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfiguration(ConnectionConfiguration configuration);
+      abstract Builder setQos(int qos);
+      abstract Builder setRetained(boolean retained);
+      abstract Write build();
+    }
+
+    /**
+     * Define MQTT connection configuration used to connect to the MQTT broker.
+     */
+    public Write withConnectionConfiguration(ConnectionConfiguration configuration) {
+      return toBuilder().setConnectionConfiguration(configuration).build();
+    }
+
+    /**
+     * Define the MQTT message quality of service.
+     *
+     * <ul>
+     * <li>Quality of Service 0 - indicates that a message should
+     * be delivered at most once (zero or one times). The message will not be persisted to disk,
+     * and will not be acknowledged across the network. This QoS is the fastest,
+     * but should only be used for messages which are not valuable - note that
+     * if the server cannot process the message (for example, there
+     * is an authorization problem), then an
+     * {@link MqttCallback#deliveryComplete(IMqttDeliveryToken)} won't be called.
+     * Also known as "fire and forget".</li>
+     *
+     * <li>Quality of Service 1 - indicates that a message should
+     * be delivered at least once (one or more times).  The message can only be delivered safely if
+     * it can be persisted on the broker.
+     * If the broker can't persist the message, the message will not be
+     * delivered in the event of a client failure.
+     * The message will be acknowledged across the network.
+     * This is the default QoS.</li>
+     *
+     * <li>Quality of Service 2 - indicates that a message should
+     * be delivered once. The message will be persisted to disk on the broker, and will
+     * be subject to a two-phase acknowledgement across the network.
+     * The message can only be delivered safely if
+     * it can be persisted on the broker.
+     * If a persistence mechanism is not specified on the broker, the message will not be
+     * delivered in the event of a client failure.</li>
+     * </ul>
+     *
+     * <p>If persistence is not configured, QoS 1 and 2 messages will still be delivered
+     * in the event of a network or server problem as the client will hold state in memory.
+     * If the MQTT client is shutdown or fails and persistence is not configured then
+     * delivery of QoS 1 and 2 messages can not be maintained as client-side state will
+     * be lost.</p>
+     *
+     * @param qos The quality of service value.
+     * @return The {@link Write} {@link PTransform} with the corresponding QoS configuration.
+     */
+    public Write withQoS(int qos) {
+      return toBuilder().setQos(qos).build();
+    }
+
+    /**
+     * Whether or not the publish message should be retained by the messaging engine.
+     * Sending a message with the retained set to {@code false} will clear the
```

Comment:
I don't understand this: do you mean that if you send a message with retained=true it will be saved, but then if you send the same message with retained=false then it will be removed from the server? How does the server identify "the same" message - do messages have some kind of id?